src/main/java/software/amazon/cloudformation/resource/Validator.java

```diff
@@ -68,7 +95,72 @@ public void validateObject(final JSONObject modelObject, final JSONObject defini
      * @throws ValidationException Thrown for any schema validation errors
      */
     public void validateResourceDefinition(final JSONObject definition) throws ValidationException {
+        // inject/replace $schema URI to ensure that provider definition schema is used
+        definition.put("$schema", RESOURCE_DEFINITION_SCHEMA_URI.toString());
         validateObject(definition, definitionSchemaJsonObject);
+        // validateObject cannot validate schema-specific attributes. For example if definition
+        // contains "propertyA": { "$ref":"./some-non-existent-location.json#definitions/PropertyX"}
+        // validateObject will succeed, because all it cares about is that "$ref" is a URI
+        // In order to validate that $ref points at an existing location in an existing document
+        // we have to "load" the schema
+        loadResourceSchema(definition);
     }
 
+    public Schema loadResourceSchema(final JSONObject resourceDefinition) {
+        return getResourceSchemaBuilder(resourceDefinition).build();
+    }
+
+    /**
+     * returns Schema.Builder with pre-loaded JSON draft-07 meta-schema and resource definition meta-schema
+     * (resource.definition.schema.v1.json). Resulting Schema.Builder can be used to build a schema that
+     * can be used to validate parts of CloudFormation template.
+     *
+     * @param resourceDefinition - actual resource definition (not resource definition schema)
+     * @return
+     */
+    public Schema.Builder<?> getResourceSchemaBuilder(final JSONObject resourceDefinition) {
+        final SchemaLoaderBuilder loaderBuilder = getSchemaLoader(resourceDefinition);
+        loaderBuilder.registerSchemaByURI(RESOURCE_DEFINITION_SCHEMA_URI, definitionSchemaJsonObject);
+
+        final SchemaLoader loader = loaderBuilder.build();
+        try {
+            return loader.load();
+        } catch (org.everit.json.schema.SchemaException e) {
+            throw new ValidationException(e.getMessage(), e.getSchemaLocation(), e);
```

Comment:
could we convert the `SchemaException` to a `org.everit.json.schema.ValidationException` and keep using `causingExceptions` instead?